### PR TITLE
Refactor tests for nonlinear solid solver

### PR DESCRIPTION
**Pull request by joshessman-llnl**
The Inlet functions PR was recently merged into axom, which allows us to define coefficients in input files (with some limitations pending a follow-up Axom PR).

This PR introduces changes to how the `NonlinearSolid` module handles information from the input file (namely, constructing Coefficients for ICs/BCs).  It also adds a new "InputInfo" struct for coefficient information (the functions).

Most of this PR is boilerplate changes, but there were a few things that might be worthy of discussion:
- How do we want to handle "variants"?  After some discussion with @white238 we decided on a string tag + subtable (see `nonlinear_solid.stiffness_solver.linear`, but possibly there's a way to add first-class support in inlet for this concept.
- Boundary condition naming: Inlet now supports dictionaries for more convenient naming of boundary conditions.  Right now I just check if the field name is present in the BC name to associate it with a field, but possibly there's a better/more readable option.

Also, I temporarily removed the `VectorScaledConstantCoefficient` from the driver pending the follow-up Axom PR, as it is effectively a time-dependent coefficient (not yet supported).

Added input doc:
https://serac.readthedocs.io/en/feature-essman-initial_inlet_functions/

### CODE REVIEW

File: data/input_files/default.lua [hunk 1]
```diff
@@ -14,12 +14,15 @@ main_mesh = {
...
     stiffness_solver = {
         linear = {
-            rel_tol     = 1.0e-6,
-            abs_tol     = 1.0e-8,
-            max_iter    = 5000,
-            print_level = 0,
-            solver_type = "gmres",
-            prec_type   = "HypreAMG",
+            type = "iterative",
+            iterative_params = {
```
Comment (jamiebramwell) [inline data/input_files/default.lua]: Depending which PR goes in first, these `params` should become `options`. The easier answer is to probably have this one go in first and then I'll fix it.

File: data/input_files/default.lua [hunk 2]
```diff
@@ -44,15 +47,40 @@ nonlinear_solid = {
         z = 0.0,
     },
 
+    -- initial conditions
+    -- initialize x_cur, boundary condition, deformation, and
+    -- incremental nodal displacment grid functions by projecting the
+    -- VectorFunctionCoefficient function onto them
+
+    initial_displacement = {
```
Comment (jamiebramwell) [inline data/input_files/default.lua:55]: 👍

File: data/input_files/tests/nonlinear_solid/dyn_amgx_solve.lua [hunk 3]
```diff
@@ -47,12 +50,31 @@ nonlinear_solid = {
...
+        end  
+    },
+
+    initial_velocity = {
+        vec_coef = function (x, y, z)
+            s = 0.1 / 64
+            first = -s * x * x
+            last = s * x * x * (8.0 - x)
+            -- FIXME: How can we detect the dimension?
+            return first, 0, last
```
Comment (jamiebramwell) [inline data/input_files/tests/nonlinear_solid/dyn_amgx_solve.lua:66]: in 2D, we typically keep 3 variables but leave the last one 0.0 (at least in older codes).

File: data/input_files/tests/nonlinear_solid/qs_attribute_solve.lua [hunk 4]
```diff
@@ -42,14 +45,20 @@ nonlinear_solid = {
 
     -- boundary condition parameters
     boundary_conds = {
-        {
-            name = "displacement_x",
+        ['displacement_x'] = {
             -- boundary attribute 1 (index 0) is fixed (Dirichlet) in the x direction
             attrs = {1},
+            component = 0,
```
Comment (jamiebramwell) [inline data/input_files/tests/nonlinear_solid/qs_attribute_solve.lua:51]: Why is the component still needed if the name of the BC is `displacement_x`?

File: tests/test_utilities.cpp [hunk 5]
```diff
@@ -30,14 +33,86 @@ void defineNonlinSolidInputFileSchema(axom::inlet::Inlet& inlet, const bool dyna
...
+  serac::NonlinearSolid::InputInfo::defineInputFileSchema(solid_solver_table);
 
   // Verify input file
   if (!inlet.verify()) {
     SLIC_ERROR("Input file failed to verify.");
   }
 }
 
-}  // end namespace testing
+void runNonlinSolidTest(const std::string& input_file)
```
Comment (jamiebramwell) [inline tests/test_utilities.cpp:44]: 🎉

File: data/input_files/tests/nonlinear_solid/dyn_amgx_solve.lua (hunk 3 above)
Comment (joshessman-llnl) [reply data/input_files/tests/nonlinear_solid/dyn_amgx_solve.lua:66]: Yep, the issue is that the current interface does not provide enough information to determine the dimension, so I've assumed 3D for the time being.  A future Axom PR might add the vector type as a first-class interface whose size can be queried.

File: data/input_files/tests/nonlinear_solid/qs_attribute_solve.lua (hunk 4 above)
Comment (joshessman-llnl) [reply data/input_files/tests/nonlinear_solid/qs_attribute_solve.lua:51]: I wasn't sure what naming convention we wanted for boundary conditions - right now I'm just checking if the name of the field is in the name of the BC and applying it accordingly.

**Review by jamiebramwell — APPROVED** [review]

File: src/serac/infrastructure/input.cpp [hunk 6]
```diff
@@ -67,24 +67,47 @@ std::string fullDirectoryFromPath(const std::string& path)
...
+{
+  auto scalar_func = std::get_if<CoefficientInputOptions::ScalarFunc>(&func);
+  SLIC_ERROR_IF(!scalar_func, "Cannot construct a scalar coefficient from a vector function");
+  return {*scalar_func};
+}
+
+void CoefficientInputOptions::defineInputFileSchema(axom::inlet::Table& table)
+{
+  table.addFunction("vec_coef", axom::inlet::FunctionType::Vec3D,
+                    {axom::inlet::FunctionType::Vec3D},  // Multiple argument types
```
Comment (white238) [inline src/serac/infrastructure/input.cpp]: are you explaining this converts automatically to x,y,z? might be better if the comment says:

`// Automatically converted to x, y, z in Lua`

File: src/serac/physics/nonlinear_solid.cpp [hunk 7]
```diff
@@ -253,38 +281,30 @@ void NonlinearSolid::advanceTimestep(double& dt)
...
-    auto& dynamics_table = table.addTable("dynamics", "Parameters for mass matrix inversion");
-    dynamics_table.addString("timestepper", "Timestepper (ODE) method to use");
-    dynamics_table.addString("enforcement_method", "Time-varying constraint enforcement method to use");
-  }
+  auto& dynamics_table = table.addTable("dynamics", "Parameters for mass matrix inversion");
+  dynamics_table.addString("timestepper", "Timestepper (ODE) method to use");
+  dynamics_table.addString("enforcement_method", "Time-varying constraint enforcement method to use");
 
-  auto& bc_table = table.addGenericArray("boundary_conds", "Boundary condition information");
+  auto& bc_table = table.addGenericDictionary("boundary_conds", "Boundary condition information");
```
Comment (white238) [inline src/serac/physics/nonlinear_solid.cpp]: Need a plural somewhere in this description.

File: src/serac/physics/utilities/equation_solver.cpp [hunk 8]
```diff
@@ -339,15 +339,34 @@ mfem::Operator& EquationSolver::SuperLUNonlinearOperatorWrapper::GetGradient(con
...
-  linear_table.addInt("print_level", "Linear print level.").defaultValue(0);
-  linear_table.addString("solver_type", "Solver type (gmres|minres).").defaultValue("gmres");
-  linear_table.addString("prec_type", "Preconditioner type (JacobiSmoother|L1JacobiSmoother|AMG|BlockILU).")
+  auto& linear_table =
+      table.addTable("linear", "Linear Equation Solver Parameters")
+          .required()
+          .registerVerifier([](const axom::inlet::Table& table) {
+            // Make sure that the provided options match the desired linear solver type
+            const bool is_iterative =
+                (table["type"].get<std::string>() == "iterative") && table.contains("iterative_options");
```
Comment (white238) [inline src/serac/physics/utilities/equation_solver.cpp:348]: 👍

File: tests/CMakeLists.txt [hunk 9]
```diff
@@ -16,13 +16,12 @@ if (ENABLE_GTEST)
         NAME        test_utils
         SOURCES     test_utilities.cpp
         HEADERS     test_utilities.hpp
-        DEPENDS_ON  serac_physics
+        DEPENDS_ON  serac_physics gtest serac_infrastructure
```
Comment (white238) [inline tests/CMakeLists.txt]: serac_infrastructure is already inherited from serac_physics

File: tests/petsc_smoketest.cpp [hunk 10]
```diff
@@ -96,7 +96,7 @@ PetscErrorCode ComputeRHS(KSP ksp, Vec b, void* ctx)
   PetscFunctionBeginUser;
   ierr = KSPGetDM(ksp, &da);
   CHKERRQ(ierr);
-  ierr = DMDAGetInfo(da, 0, &M, &N, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0);
+  ierr = DMDAGetoptions(da, 0, &M, &N, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0);
```
Comment (white238) [inline tests/petsc_smoketest.cpp]: DMDAGetOptions

Comment (white238) [reply tests/petsc_smoketest.cpp]: Nevermind I guess this isn't ours or if it is I'm too tired to find it.

**Review by white238 — APPROVED** [review]

File: tests/petsc_smoketest.cpp (hunk 10 above)
Comment (jamiebramwell) [reply tests/petsc_smoketest.cpp]: Yep, it's not ours.

**Comment by joshessman-llnl** [comment]
LGTM
